### DISCUSSION TITLE
Support using Chakra with Styled-Components

### DISCUSSION THREAD

**Discussion by ghost** [chakra-ui/chakra-ui#3435] [Ideas]
I tried this but it didn't work, `CSS in JS` gets really messy that's why I like `styled-components`.

```typescript
import { Button } from "@chakra-ui/react"

class FooBar extends Component {
    render() {
        return <StyledButton onClick={e => console.log('clicked')} />;
    }
}

const StyledButton = styled(Button)`
    background: black;
`
```

Is it because of issue #2338 about missing adapter for styled-components. Can't you just pass styled props to the component like its overriding CSS? Rebass supports both styled-components and css-in-js with styled system.

`styled-components` vs. `CSS in JS` are two really seperate ways of styling stuff that makes it a big barrier for using `chakra` if developers have to learn completely new ways of doing CSS, arguably one better than the other sure but that doesn't help chakra-ui if you're forcing developers to adopt a different syntax

**Comment by anilanar** [comment]
@bachmannn I thought chakra-ui used emotion which has almost identical API to styled components. Am I mistaken?

**Reply by with-heart** [reply]
We use `emotion` under the hood but don't support the tagged template syntax offered by `@emotion/styled`. It was a tradeoff for enabling our theme-driven style props system.

**Comment by dodas** [comment]
Hi @bachmannn,

using `styled-components` with Chakra UI is not supported and probably will never be.
However, porting code that is using `styled-components` to `emotion`/`chakra` should be pretty trivial.
You will need to convert your styles from string syntax to object syntax, and use `chakra` factory in place of `styled`:
```jsx
import { chakra, Button } from "@chakra-ui/react"

const StyledButton = chakra(Button, {
  baseStyle: {
    background: "black",
  },
})
```

[Read more about `chakra` factory](https://chakra-ui.com/docs/features/chakra-factory)

If you don't need to reuse the `StyledButton` component, I'd prefer just passing style props to `Button`:
```jx
<Button background="black">button</Button>
```

**Reply by jansengunderson** [reply]
I'm finding that adding a sizes object does not work when using chakra factory. Only baseStyle is passed to the component. How else could one add responsive styles?

<pre>
const StatNumber = chakra(Heading, {
    baseStyle: {
        fontSize: "6rem",
        textAlign: "center",
    },
    sizes: {
        md: {
            fontSize: '2rem'
        },
        lg: {
            fontSize: '4.5rem'
        },
    }
});
</pre>

**Reply by johkker** [reply]
> I'm finding that adding a sizes object does not work when using chakra factory. Only baseStyle is passed to the component. How else could one add responsive styles?
> 
> const StatNumber = chakra(Heading, {
>     baseStyle: {
>         fontSize: "6rem",
>         textAlign: "center",
>     },
>     sizes: {
>         md: {
>             fontSize: '2rem'
>         },
>         lg: {
>             fontSize: '4.5rem'
>         },
>     }
> });

You could probably use fontSizes instead of fontSize and set your breakpoints, but I don't think it would be needed, cause the main theme already has those breakpoints. 

It would be something like:

```
const StatNumber = chakra(Heading, {
    baseStyle: {
        fontSizes: {
            md: "2rem",
            lg: "4.5rem",
          },
        textAlign: "center",
    },

});
```

**Comment by corbanbrook** [comment]
This is a little bit of a hack but If you want to use Chakra with a styled-component i found it helpful to create an `sx` util when you need to access the chakra props and theme vars:

```typescript
import { css, Theme, StyleObjectOrFn } from '@chakra-ui/react'

export const sx = (styles: StyleObjectOrFn) => (props: { theme: Theme }) => css(styles)(props.theme)
```

```typescript
const StyledThing = styled.div`
  ${sx({ textStyle: 'body', color: 'black.50' )}
  
  // mixed with some additional css in the template literal
  transition: opacity 200ms ease-out;
  opacity: 0.5;

  &:hover {
     opacity: 1;
  }
`
```

**Reply by maikenegreiros** [reply]
How could I work with pseudo elements in this object style?

**Reply by richardaum** [reply]
> How could I work with pseudo elements in this object style?

https://chakra-ui.com/docs/styled-system/style-props#pseudo

**Reply by JSONRice** [reply]
This is confusing.

**Comment by picninim** [comment]
If you use it with Typescript you can still use StyledComponents + Chakra and keep the IDE functionalities

```typescript
import { Button, ButtonProps } from "@chakra-ui/react";
import styled from "styled-components";

export const MyButton = styled(Button).attrs({
    colorScheme: "gray",
} as ButtonProps)`
 // some other css
` as typeof Button;
```
Use it normally and still get access to extend it
```jsx
<MyButton onClick={handleClickFn}> </MyButton>
```

You can also access globalHooks like `useColorModeValue()` as long you pass the attrs as a Fn
```typescript
import { Button, ButtonProps, useColorModeValue } from "@chakra-ui/react";
import styled from "styled-components";

export const MyButton = styled(Button).attrs(() => ({
    bg: useColorModeValue("gray.400", "gray.700"),
} as as ButtonProps ))`
 // some other css
` as typeof Button;
```

You need to pass `as ButtonProps` as `as typeof Button` so the IDE can get the right interface of the ChakraUI and not the StyledComponent

**Reply by TangoPJ** [reply]
I noticed that even if we can delete text "// some other css", we should leave an empty string, otherwise there will be an error

**Reply by TangoPJ** [reply]
How can I provide screen sizes to change styles in this case?

**Reply by picninim** [reply]
> I noticed that even if we can delete text "// some other css", we should leave an empty string, otherwise there will be an error

Yes, you need to keep an empty string if don't plan to add direct css, once that is the template string necessary for the styled component.
You can still provide media queries in the styled component css or using chakra props:

In styled component:

``` typescript
export const MyButton = styled(Button).attrs(() => ({
    bg: useColorModeValue("gray.400", "gray.700"),
} as as ButtonProps ))`
  @media (max-width: 768px) {
    position: absolute;
  }
` as typeof Button;
```

Or in charka props:
``` jsx
<MyButton position={{ base: 'initial', md: 'absolute' }}>
```

**Comment by sim0nides** [comment]
You can use combination of `@emotion/styled` what is core of Chakra styling and `css` function so you don't have to install any additional packages.

```
import { css, Td } from '@chakra-ui/react'
import styled from '@emotion/styled'

const Cell = styled(Td)(props => (
    css({
        bg: props.success ? 'green.300' : 'red.300'
    })(props.theme)
))

<Cell success={false} />
```

**Reply by Ryusse** [reply]
In this case how should I use the media query of the chakraui theme?
